packages/browser/core/src/core/auth/internal/StitchAuthImpl.ts

```diff
@@ -140,12 +265,108 @@ export default class StitchAuthImpl extends CoreStitchAuth<StitchUser>
       });
     }
   }
+
+  public setRedirectFragment(key: RedirectKeys | RedirectFragmentFields, value: any) {
+    this.storageRef.set(key, value);
+  }
+}
+
+function generateState(): string {
+  let state = "";
+  for (let i = 0; i < 64; ++i) {
+    state += alphaNumericCharacters.charAt(Math.floor(Math.random() * alphaNumericCharacters.length));
+  }
+
+  return state;
+}
+
+export function prepareRedirect(
+  auth: StitchAuthImpl,
+  credential: StitchRedirectCredential,
+  redirectType: RedirectTypes
+): { redirectUrl: string; state: string } {
+  auth.setRedirectFragment(RedirectKeys.RedirectProvider, credential.providerType);
+  let redirectUrl = credential.redirectUrl;
+  if (redirectUrl === undefined) {
+    redirectUrl = pageRootUrl();
+  }
+
+  const state = generateState();
+  auth.setRedirectFragment(RedirectFragmentFields.State, state);
+  auth.setRedirectFragment(RedirectKeys.Type, redirectType);
+
+  return { redirectUrl, state };
+}
+
+function pageRootUrl(): string {
+  return [
+    window.location.protocol,
+    "//",
+    window.location.host,
+    window.location.pathname
+  ].join("");
+}
+
+function unmarshallUserAuth(data): AuthInfo {
+  const parts = data.split("$");
+  if (parts.length !== 4) {
+    throw new RangeError("invalid user auth data provided: " + data);
```

Comment:
make this a stitch error with more info relative to the request

Reply:
👍